### PR TITLE
API: Timestamp and Timedelta .value changing in 2.0

### LINKED ISSUE #49076: API: Timestamp and Timedelta .value changing in 2.0
```python
import pandas as pd
import numpy as np

dt = np.datetime64("2016-01-01", "ms")
ts = pd.Timestamp(dt)

>>> ts.value
1451606400000        # <- 2.0/main
1451606400000000000  # <- 1.x
```

In previous versions .value has always been in nanoseconds.  By changing the reso we get with some Timestamp/Timedelta inputs, we change the .value, which is technically public API.

One option would be just to document the change in .value behavior along with the other breaking changes docs.

Another would be to use e.g. `._value` internally and keep `.value` as representing nanos.

### PR DESCRIPTION
**Pull request by MarcoGorelli**
- [ ] closes #49076 (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
failed test:
```
FAILED pandas/tests/io/json/test_pandas.py::TestPandasContainer::test_date_index_and_values[date-True-epoch] - assert '{"1577836800...null,"a":"a"}' == '{"1577836800...null,"a":"a"}'
  - {"1577836800000":1577836800000,"null":null,"a":"a"}
  ?                                 ^^^^
  + {"1577836800000":1577836800000,"-6858695778871":null,"a":"a"}
  ?   
```

can't get my head round it - cc @jbrockmendel @WillAyd in case you have any ideas, else I'll keep at it

renaming `.value` affects _so_ much of the codebase..

**Comment by jbrockmendel** [comment]
> can't get my head round it - cc @jbrockmendel @WillAyd in case you have any ideas, else I'll keep at it

best guess is in objToJSON.c there are a couple of `get_long_attr(item, "value")` calls that might not be caught by grepping for `.value`

**Comment by WillAyd** [comment]
I think the JSON module returns garbage values because of https://github.com/pandas-dev/pandas/issues/49756

If you look, there are some branches in the code that do something like: 

```c
            if (PyObject_HasAttrString(item, "value")) {
                // see test_date_index_and_values for case with non-nano
                nanosecVal = get_long_attr(item, "value");
            } else {
                if (PyDelta_Check(item)) {
                    nanosecVal = total_seconds(item) *
                                 1000000000LL;  // nanoseconds per second
                } else {
                    // datetime.* objects don't follow above rules
                    nanosecVal = PyDateTimeToEpoch(item, NPY_FR_ns);
                }
            }
```

For the timedelta, this goes into `total_seconds` which is implemented as such:

```c
static npy_float64 total_seconds(PyObject *td) {
    npy_float64 double_val;
    PyObject *value = PyObject_CallMethod(td, "total_seconds", NULL);
    double_val = PyFloat_AS_DOUBLE(value);
    Py_DECREF(value);
    return double_val;
}
```

I'm guessing that the object doesn't actually have `total_seconds` defined, and you end up getting pretty crazy results. 

cc @lithomas1 who may be interested

**Comment by WillAyd** [comment]
As the master of code checks @MarcoGorelli if you had some kind of idea on how to set up CI so that all `result = Py...` C functions are immediately followed by an `if (result == NULL) { // handle error }` that could help us clean up our extensions. Otherwise its probably worth doing a manual review of things as a pre-cursor to this PR

**Review by MarcoGorelli — COMMENTED** [review]
gonna block myself here, want to make sure this is really what's desired, as then the public method `.value` would become unavailable for old dates: https://github.com/pandas-dev/pandas/issues/49076#issuecomment-1408481395

**Comment by MarcoGorelli** [comment]
> gonna block myself here, want to make sure this is really what's desired, as then the public method `.value` would become unavailable for old dates: [#49076 (comment)](https://github.com/pandas-dev/pandas/issues/49076#issuecomment-1408481395)

have amended the error message to suggest using `asm8` if nanoseconds aren't what they're after

File: pandas/_libs/tslibs/nattype.pyi [hunk 1]
```diff
@@ -19,6 +19,7 @@ class _NatComparison:
 
 class NaTType:
     value: np.int64
+    _value: np.int64
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/nattype.pyi:22]: did you add a .value property for NaT?

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/nattype.pyi:22]: good point, thanks - should it have the same value, right?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/nattype.pyi:22]: NPY_NAT, yes

Comment (MarcoGorelli) [reply pandas/_libs/tslibs/nattype.pyi:21]: cool, have added, thanks

any further comments? on board with the change, or should something else be used in place of `._value`?

Comment (jbrockmendel) [reply pandas/_libs/tslibs/nattype.pyi:21]: _value works for me

File: pandas/_libs/tslibs/timedeltas.pyi [hunk 2]
```diff
@@ -89,6 +89,7 @@ class Timedelta(timedelta):
     max: ClassVar[Timedelta]
     resolution: ClassVar[Timedelta]
     value: int  # np.int64
+    _value: int  # np.int64
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/timedeltas.pyi:92]: to be Technically Correct do we need to make value a property here?

File: pandas/core/arrays/_ranges.py [hunk 3]
```diff
@@ -46,8 +46,8 @@ def generate_regular_range(
     ndarray[np.int64]
         Representing the given resolution.
     """
-    istart = start.value if start is not None else None
-    iend = end.value if end is not None else None
+    istart = start._value if start is not None else None
```
Comment (jbrockmendel) [inline pandas/core/arrays/_ranges.py:49]: im a bit concerned about this.  we have a unit keyword that i think is implicitly assuming unit == start.unit == end.unit

Comment (jbrockmendel) [reply pandas/core/arrays/_ranges.py:49]: no wait, never mind, this is doing the right thing given that concern.

File: pandas/core/dtypes/cast.py [hunk 4]
```diff
@@ -814,7 +814,7 @@ def infer_dtype_from_scalar(val, pandas_dtype: bool = False) -> tuple[DtypeObj,
             dtype = _dtype_obj
         else:
             dtype = np.dtype("m8[ns]")
-            val = np.timedelta64(val.value, "ns")
+            val = np.timedelta64(val._value, "ns")
```
Comment (jbrockmendel) [inline pandas/core/dtypes/cast.py:817]: i think this one is expecting the nanos value?  (and probably needs to be updated to allow non-nano?)

File: pandas/core/indexes/interval.py [hunk 5]
```diff
@@ -546,7 +546,7 @@ def _maybe_convert_i8(self, key):
             if lib.is_period(key):
                 key_i8 = key.ordinal
             elif isinstance(key_i8, Timestamp):
-                key_i8 = key_i8.value
+                key_i8 = key_i8._value
```
Comment (jbrockmendel) [inline pandas/core/indexes/interval.py:549]: looks like 1) this scalar path might not be necessary? and 2) there may be baked-in assumptions about everything being nano

Comment (MarcoGorelli) [reply pandas/core/indexes/interval.py:549]: good call - shall we make logic changes in a separate PR? I've opened https://github.com/pandas-dev/pandas/issues/51196 about this nanosecond assumption here

Comment (jbrockmendel) [reply pandas/core/indexes/interval.py:549]: separate PR seems fine

Comment (MarcoGorelli) [reply pandas/core/indexes/interval.py:549]: sure, thanks - any objections here? shall we move forward before merge conflicts arise?

**Review by mroeschke — COMMENTED** [review]
Looks fairly good. Could this use a whatsnew note about `.values` raising a OverflowError for non nano unit?

**Comment by MarcoGorelli** [comment]
thanks - @jbrockmendel can confirm but I don't think this (non-nano timestamp) would've been available anyway in 1.5.x

**Comment by jbrockmendel** [comment]
> I don't think this (non-nano timestamp) would've been available anyway in 1.5.x

correct

**Review by mroeschke — APPROVED** [review]

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Comment by jbrockmendel** [comment]
thanks @MarcoGorelli

**Pull request merged**